### PR TITLE
ContentHash integration for File/DataLake

### PR DESCRIPTION
**Pull request by katmsft**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp [hunk 1]
```diff
@@ -64,7 +50,25 @@ namespace Azure { namespace Storage {
     /**
      * @brief The algorithm used for hash.
      */
-    HashAlgorithm Algorithm;
+    HashAlgorithm Algorithm = HashAlgorithm::Md5;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:53]: I was thinking this question. Maybe it's not the best idea to define this value here.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:53]: We should have an `HttpHeaders` struct in common package, because in all three services, the definition of this struct is the same. We can add a default constructor, where we init the default value of member `ContentHash` to `Md5`.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp [hunk 2]
```diff
@@ -64,7 +50,25 @@ namespace Azure { namespace Storage {
...
+      {
+        return std::lexicographical_compare(
+            lhs.begin(), lhs.end(), rhs.begin(), rhs.end(), [](char c1, char c2) {
+              return Core::Strings::ToLower(c1) < Core::Strings::ToLower(c2);
+            });
+      }
+    };
+
+    ContentHash FromBase64String(const std::string& base64String, HashAlgorithm algorithm);
+    std::string ToBase64String(const ContentHash& hash);
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:70]: The interface of these two functions are not symmetrical, but their names are symmetrical.

In the first function you have input (base64 encoded string, algorithm) and the output is `ContentHash`.
But in the second one, the output is just base64 encoded string, the algorithm information is lost.

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,8 @@
 
 - Move DataLake SAS into `Azure::Storage::Sas` namespace.
 - `EncrytionKeySha256` are changed to binary(`std::vector<uint8_t>`).
+- Replaced all transactional content MD5/CRC64 with `ContentHash` struct.
+- `DataLakeHttpHeaders` is renamed to `PathHttpHeaders`, and now contains `ContentHash` for the resource.
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/CHANGELOG.md:10]: `PathHttpHeaders`? This doesn't sound relevant to this PR.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp [hunk 4]
```diff
@@ -2203,8 +2201,9 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
               ContentLength; // Required for "Append Data" and "Flush Data".  Must be 0 for "Flush
                              // Data".  Must be the length of the request content in bytes for
                              // "Append Data".
-          Azure::Core::Nullable<std::string> ContentMd5; // Specify the transactional md5 for the
-                                                         // body, to be validated by the service.
+          Azure::Core::Nullable<Storage::ContentHash>
+              ContentMd5; // Specify the transactional md5 for the body, to be validated by the
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2205]: Why not `TransactionalContentHash`?

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp [hunk 5]
```diff
@@ -2365,12 +2366,12 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
...
-          Azure::Core::Nullable<std::string>
+          Azure::Core::Nullable<Storage::ContentHash>
               TransactionalContentMd5; // Specify the transactional md5 for the body, to be
                                        // validated by the service.
-          Azure::Core::Nullable<std::string>
-              ContentCrc64; // Specify the transactional crc64 for the body, to be validated by the
-                            // service.
+          Azure::Core::Nullable<Storage::ContentHash>
+              TransactionalContentCrc64; // Specify the transactional crc64 for the body, to be
+                                         // validated by the service.
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2374]: Why two separate hashes?

File: sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp [hunk 6]
```diff
@@ -58,4 +58,22 @@ namespace Azure { namespace Storage {
...
+  /**
+   * @brief Specifies HTTP options for conditional requests based on ContentHash.
+   */
+  struct ContentHashAccessConditions
+  {
+    /**
+     * @brief Specify this header to perform the operation only if the resource's ContentHash
+     * matches the value specified.
+     */
+    Azure::Core::Nullable<ContentHash> IfMatch;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp:70]: `IfMatch` and `IfNoneMatch` is too confusing

File: sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp [hunk 7]
```diff
@@ -230,7 +240,12 @@ namespace Azure { namespace Storage { namespace Files { namespace DataLake {
     protocolLayerOptions.RetainUncommittedData = options.RetainUncommittedData;
     protocolLayerOptions.Close = options.Close;
     protocolLayerOptions.ContentLength = 0;
-    protocolLayerOptions.ContentMd5 = options.ContentMd5;
+    if (options.ContentHash.HasValue()
+        && options.ContentHash.GetValue().Algorithm == HashAlgorithm::Crc64)
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-datalake/src/datalake_file_client.cpp:244]:
```suggestion
        && options.ContentHash.GetValue().Algorithm != HashAlgorithm::Md5)
```

File: sdk/storage/azure-storage-files-shares/CHANGELOG.md [hunk 8]
```diff
@@ -12,6 +12,8 @@
   - ShareClient::BreakLease
   - ShareClient::RenewLease
 - Move File SAS into `Azure::Storage::Sas` namespace.
+- Replaced all transactional content MD5/CRC64 with `ContentHash` struct.
+- `FileShareHttpHeaders` is renamed to `ShareFileHttpHeaders`, and member `std::string ContentMd5` is changed to `Storage::ContentHash ContentHash`.
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/CHANGELOG.md:16]: Is this change relevant to this PR?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 9]
```diff
@@ -5566,7 +5567,12 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
           Azure::Core::Nullable<std::string>
               FileCacheControl; // Sets the file's cache control. The File service stores this value
                                 // but does not use or modify it.
-          Azure::Core::Nullable<std::string> FileContentMd5; // Sets the file's MD5 hash.
+          Azure::Core::Nullable<Storage::ContentHash>
+              ContentMd5; // An MD5 hash of the content. This hash is used to verify the integrity
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:5571]: `ContentHash`?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 10]
```diff
@@ -5828,7 +5836,12 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:5840]: Is this transactional or not?

It should be named `TransactionalContentHash` if it's transactional, otherwise, `ContentHash`.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:5840]: ditto for the others.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 11]
```diff
@@ -6289,13 +6305,13 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
+          Azure::Core::Nullable<Storage::ContentHash>
               SourceContentCrc64; // Specify the crc64 calculated for the range of bytes that must
                                   // be read from the copy source.
-          Azure::Core::Nullable<std::string>
+          Azure::Core::Nullable<Storage::ContentHash>
               SourceIfMatchCrc64; // Specify the crc64 value to operate only on range with a
                                   // matching crc64 checksum.
-          Azure::Core::Nullable<std::string>
+          Azure::Core::Nullable<Storage::ContentHash>
               SourceIfNoneMatchCrc64; // Specify the crc64 value to operate only on range without a
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:6315]: I have no idea how we should name these variables.

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:6315]: Let's discuss offline.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_options.hpp [hunk 12]
```diff
@@ -662,22 +662,17 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
      */
-    Azure::Core::Nullable<std::string> SourceContentCrc64;
+    Azure::Core::Nullable<ContentHash> SourceContentHash;
 
     /**
-     * @brief Specify the crc64 value to operate only on range with a matching crc64 checksum.
+     * @brief Specify the access condition for the source. Only Crc64 is supported.
      */
-    Azure::Core::Nullable<std::string> SourceIfMatchCrc64;
+    ContentHashAccessConditions SourceContentHashAccessCondition;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_options.hpp:670]: `SourceAccessConditions`?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_options.hpp [hunk 13]
```diff
@@ -662,22 +662,17 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
-     * @brief Specify the crc64 value to operate only on range without a matching crc64 checksum.
+     * @brief The operation will only succeed if the lease access condition is met.
      */
-    Azure::Core::Nullable<std::string> SourceIfNoneMatchCrc64;
-
-    /**
-     * @brief The operation will only succeed if the access condition is met.
-     */
-    LeaseAccessConditions AccessConditions;
+    Storage::LeaseAccessConditions LeaseAccessConditions;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_options.hpp:675]: `AccessConditions`?

Naming it `LeaseAccessConditions` will cause that versioning issue. What if in the future we support more access conditions?

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp [hunk 14]
```diff
@@ -853,9 +862,13 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
     {
       protocolLayerOptions.FileContentDisposition = options.HttpHeaders.ContentDisposition;
     }
-    if (!options.HttpHeaders.ContentMd5.empty())
+    if (!options.HttpHeaders.ContentHash.Value.empty())
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:865]: How should we denote empty hash?

Now the algorithm is Md5, while the value is empty. This looks kind of weird.

**Review by Jinming-Hu — CHANGES_REQUESTED** [review]
See in comments

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:53]: This is not how .NET does it.
I'm wondering if it is for versioning purposes, to prevent breaking changes if one HttpHeaders needs to contain more than what it has at the moment.
We can check with .NET team or Ted or Jeffrey. However, I don't think we should make the change before we have this change approved.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:70]: How about `FromBase64StringHash`? I don't have good namings in mind.

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/CHANGELOG.md:10]: This is relevant for this RP due to alignment to .NET SDK for HttpHeaders, including the name of the HttpHeaders and it's content.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2205]: https://github.com/Azure/azure-sdk-for-cpp/issues/1237.
Opened an issue to track this, the doc is wrong, not the type itself.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp (hunk 5 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2374]: I don't think we reached a conclusion that the non-user-facing interface has the requirement of unifying content hash for both CRC64 and Md5.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp:70]: Any proposals?
How about `IfHashMatch` and `IfNoneHashMatch`?

File: sdk/storage/azure-storage-files-shares/CHANGELOG.md (hunk 8 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/CHANGELOG.md:16]: This is relevant for this RP due to alignment to .NET SDK for HttpHeaders, including the name of the HttpHeaders and it's content.

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp (hunk 14 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:865]: I can accept it since there is no way an empty value Md5 hash being valid.

File: sdk/storage/azure-storage-files-datalake/CHANGELOG.md (hunk 3 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/CHANGELOG.md:10]: Fine

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp (hunk 4 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2205]: The variable name is also wrong

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp (hunk 5 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2374]: I can live with that. But I still think it's far from satisfactory.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp (hunk 6 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp:70]: `IfMatchMatch` and `IfHashNoneMatch`?

or

`IfContentHashMatch` and `IfContentHashNoneMatch`?

File: sdk/storage/azure-storage-files-shares/CHANGELOG.md (hunk 8 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/CHANGELOG.md:16]: ok

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp (hunk 14 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:865]: Should we add another enum `None` besides `Md5` and `Crc64`?

**Comment by Jinming-Hu** [comment]
Please make sure you've added precondition check for all `HttpHeaders.ContentHash.Algorithm` being `Md5`.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp:70]: `IfContentHashMatch` and `IfContentHashNoneMatch` are better. Will make those changes.

File: sdk/storage/azure-storage-files-shares/src/share_file_client.cpp (hunk 14 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/src/share_file_client.cpp:865]: To align with other enums, we should be naming it `Unknown`.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/access_conditions.hpp:70]: @JeffreyRichter @tg-msft Please raise your concern if any.

File: sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp:2205]: The variable name is not visible to customers.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:53]: Talked offline, will remain as is for now.

File: sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-common/inc/azure/storage/common/storage_common.hpp:70]: Will remain as is before we had a good name.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 9 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:5571]: The variable name is not visible to customers.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 10 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:5840]: The variable name is not visible to customers.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 11 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:6315]: Discussed offline, and those names are not visible for customers.

**Review by Jinming-Hu — APPROVED** [review]
